src/System.Net.NameResolution/src/System/Net/NameResolutionPal.Unix.cs

```diff
@@ -0,0 +1,246 @@
+                    {
+                        continue;
+                    }
+
+                    numAddresses++;
+                }
+
+                if (numAddresses == 0)
+                {
+                    ipAddresses = Array.Empty<IPAddress>();
+                }
+                else
+                {
+                    ipAddresses = new IPAddress[numAddresses];
+                    Interop.libc.addrinfo* ai = addrinfo;
+                    for (int i = 0; i < numAddresses; ai = ai->ai_next)
+                    {
+                        Debug.Assert(ai != null);
+
+                        if ((ai->ai_family != Interop.libc.AF_INET) &&
+                            (ai->ai_family != Interop.libc.AF_INET6 || !SocketProtocolSupportPal.OSSupportsIPv6))
+                        {
+                            continue;
+                        }
+
+                        var sockaddr = new SocketAddress(
+                            ai->ai_family == Interop.libc.AF_INET ? AddressFamily.InterNetwork : AddressFamily.InterNetworkV6,
+                            checked((int)ai->ai_addrlen));
+                        for (int d = 0; d < ai->ai_addrlen; d++)
+                        {
+                            sockaddr[d] = ((byte*)ai->ai_addr)[d];
+                        }
+
+                        if (ai->ai_family == Interop.libc.AF_INET)
+                        {
+                            ipAddresses[i] = ((IPEndPoint)IPEndPointStatics.Any.Create(sockaddr)).Address;
+                        }
+                        else
+                        {
+                            ipAddresses[i] = ((IPEndPoint)IPEndPointStatics.IPv6Any.Create(sockaddr)).Address;
+                        }
+
+                        i++;
+                    }
+                }
+            }
+            finally
+            {
+                if (addrinfo != null)
+                {
+                    Interop.libc.freeaddrinfo(addrinfo);
+                }
+            }
+
+            hostinfo = new IPHostEntry
+            {
+                HostName = canonicalName ?? name,
+                Aliases = Array.Empty<string>(),
+                AddressList = ipAddresses
+            };
+            nativeErrorCode = 0;
+            return SocketError.Success;
+        }
+
+        public static unsafe string TryGetNameInfo(IPAddress addr, out SocketError socketError, out int nativeErrorCode)
+        {
+            SocketAddress address = (new IPEndPoint(addr, 0)).Serialize();
+            StringBuilder hostname = new StringBuilder(Interop.libc.NI_MAXHOST);
+
+            // TODO: Remove the copying step to improve performance. This requires a change in the contracts.
+            byte[] addressBuffer = new byte[address.Size];
+            for (int i = 0; i < address.Size; i++)
+            {
+                addressBuffer[i] = address[i];
+            }
```

Comment:
That's actually not possible: we're accessing the data through the `SocketAddress` indexer and `BlockCopy` requires an `Array` type. I am tracking the TODO with #2891 (will update the comment shortly).